From 4e4dc8c62886ca5647ec576e878a6d3f2b06132f Mon Sep 17 00:00:00 2001
From: wertherzhang <wertherzhang@china-liantong.com>
Date: Fri, 22 Nov 2013 15:38:37 +0800
Subject: [PATCH] add path to window name but the max length is 32

---
 names.c |   16 ++++++++++++++--
 1 file changed, 14 insertions(+), 2 deletions(-)

diff --git a/names.c b/names.c
index f536d2f..2cc4d45 100644
--- a/names.c
+++ b/names.c
@@ -48,6 +48,8 @@ window_name_callback(unused int fd, unused short events, void *data)
 {
 	struct window	*w = data;
 	char		*name, *wname;
+    char *path = NULL, *wpath = NULL;
+    char window_name[16];
 
 	if (w->active == NULL)
 		return;
@@ -85,11 +87,21 @@ window_name_callback(unused int fd, unused short events, void *data)
 		wname = name;
 	}
 
-	if (strcmp(wname, w->name)) {
-		window_set_name(w, wname);
+    path = osdep_get_cwd(w->active->fd);
+    if (path != NULL) {
+        wpath = parse_window_name(path);
+    }
+
+    if (wpath != NULL) {
+        snprintf(window_name, sizeof(window_name), "%s-%s", wname, wpath);
+    }
+
+	if (strcmp(window_name, w->name)) {
+		window_set_name(w, window_name);
 		server_status_window(w);
 	}
 	free(wname);
+    free(wpath);
 }
 
 char *
-- 
1.7.9.5

